bftclient/src/bft_client.cpp

```diff
@@ -12,28 +12,71 @@
-//
-// Since both of these are just instances of a `ClientRequestMsgHeader` followed by the message
-// data, we construct them here, rather than relying on the type constructors embedded into the
-// bftEngine impl. This allows us to not have to link with the bftengine library, and also allows us
-// to return the messages as vectors with proper RAII based memory management.
-Msg makeClientMsg(const RequestConfig& config, Msg&& request, bool read_only, uint16_t client_id) {
+Client::Client(std::unique_ptr<bft::communication::ICommunication> comm, const ClientConfig& config)
+    : communication_(std::move(comm)),
+      config_(config),
+      quorum_converter_(config_.all_replicas, config_.f_val, config_.c_val),
+      expected_commit_time_ms_(config_.retry_timeout_config.initial_retry_timeout.count(),
+                               config_.retry_timeout_config.number_of_standard_deviations_to_tolerate,
+                               config_.retry_timeout_config.max_retry_timeout.count(),
+                               config_.retry_timeout_config.min_retry_timeout.count(),
+                               config_.retry_timeout_config.samples_per_evaluation,
+                               config_.retry_timeout_config.samples_until_reset,
+                               config_.retry_timeout_config.max_increasing_factor,
+                               config_.retry_timeout_config.max_decreasing_factor),
+      metrics_(config.id),
+      histograms_(std::unique_ptr<Recorders>(new Recorders(config.id))) {
+  // secrets_manager_config can be set only if transaction_signing_private_key_file_path is set
+  if (config.secrets_manager_config) ConcordAssert(config.transaction_signing_private_key_file_path != std::nullopt);
+  if (config.transaction_signing_private_key_file_path) {
+    // transaction signing is enabled
+    auto file_path = config.transaction_signing_private_key_file_path.value();
+    std::optional<std::string> key_plaintext;
+
+    if (config.secrets_manager_config) {
+      // private key file is encrypted, to decrypt it we need to use the secrets manager enc
+      SecretsManagerEnc sme(config.secrets_manager_config.value());
+      key_plaintext = sme.decryptFile(file_path);
+    } else {
+      // private key file is in plain text, use secrets manager plain to read the file
+      SecretsManagerPlain smp;
+      key_plaintext = smp.decryptFile(file_path);
+    }
+    if (!key_plaintext) throw InvalidPrivateKeyException(file_path, config.secrets_manager_config != std::nullopt);
+    transaction_signer_ = RSASigner(key_plaintext.value());
+  }
+  communication_->setReceiver(config_.id.val, &receiver_);
+  communication_->Start();
+}
+
+Msg Client::makeClientMsg(const RequestConfig& config, Msg&& request, bool read_only, uint16_t client_id) {
   uint8_t flags = read_only ? READ_ONLY_REQ : EMPTY_FLAGS_REQ;
+  size_t expected_sig_len = 0;
   if (config.pre_execute) {
     flags |= PRE_PROCESS_REQ;
```

Comment:
Looks like a bug in the existing code. READ_ONLY_REQ could not be marked also PRE_PROCESS_REQ. Please fix.

Reply:
Done

Reply:
This was done on purpose. Read-only requests skip consensus entirely and go directly to individual replicas.

When and why would you issue a read only request for pre-execution?

Reply:
I misread the comment/. code. It's good as is.